synapse/storage/database.py

```diff
@@ -536,6 +538,31 @@ def new_transaction(
             **kwargs
         """
 
+        # Robustness check: ensure that none of the arguments are generators, since that
+        # will fail if we have to repeat the transaction.
+        # For now, we just log an error, and hope that it works on the first attempt.
+        # TODO: raise an exception.
+        for arg in args:
+            if inspect.isgenerator(arg):
+                logger.error("Programming error: generator passed to new_transaction")
+        for name, val in kwargs.items():
+            if inspect.isgenerator(val):
+                logger.error(
+                    "Programming error: generator passed to new_transaction as argument %s",
+                    name,
+                )
+        # also check variables referenced in func's closure
+        if inspect.isfunction(func):
+            f = cast(types.FunctionType, func)
+            if f.__closure__:
+                for i, cell in enumerate(f.__closure__):
+                    if inspect.isgenerator(cell.cell_contents):
+                        logger.error(
+                            "Programming error: function %s references generator %s via its closure",
+                            f,
+                            f.__code__.co_freevars[i],
```

Comment:
Should this be `co_cellvars` instead of `co_freevars`? https://docs.python.org/3/reference/datamodel.html#index-55 is a bit hard to follow...

Reply:
empirically not.

I've not been able to find a good reference for this, either. In short, suppose you have:

```python
def foo():
    x=1
    def bar():
         print(x)
```

then, for `foo`, `x` is a `cellvar`, but for `bar`, it is a `freevar`.

```
>>> def foo():
...     x=1
...     def bar():
...          print(x)
...     print(f"bar: {bar.__code__.co_cellvars}, {bar.__code__.co_freevars}")
... 
>>> print(f"foo: {foo.__code__.co_cellvars}, {foo.__code__.co_freevars}")
foo: ('x',), ()
>>> foo()
bar: (), ('x',)
```

Reply:
Thanks for the example, that's enlightening! Re-reading the docs again it seems like this can only refer to variables from a closure, so I guess the ordering and length must be equal.